### PR TITLE
f-statistics@v0.4.0 - Ability to provide properties during initialisation

### PR DESCRIPTION
**Pull request by tommcclean**
We want to add properties during initialisation from the web host, which will go up with every stat. For things like experiment / feature buckets, or code versions.

Non breaking and optional.

https://github.com/tommcclean/fozzie-components/blob/f-statistics%40v0.4.0/packages/services/f-statistics/README.md

```
v0.4.0
------------------------------
*October 19, 2021*

### Added
- Ability to pass in custom properties during initialization which stick around
- Above properties are now published with every statistic
```

### CODE REVIEW

File: packages/services/f-statistics/src/tests/index.test.js [hunk 1]
```diff
@@ -6,8 +7,118 @@ describe('f-statistics', () => {
         expect(StatisticsClient).toBeDefined();
     });
 
-    it('should define expected method', () => {
+    describe('constructor ::', () => {
+        it('should expose base payload when none is provided', () => {
+            // Arrange
+            const expectedPayload = {
+                je_feature: 'f-statistics',
```
Comment (xander-marjoram) [inline packages/services/f-statistics/src/tests/index.test.js]: I'd maybe pull this out into a `baseExpectedPayload` and then for some of the later tests you can do stuff like this:
```javascript
const expectedPayload = {
    ...baseExpectedPayload,
    a_test_property: 'this is a test'
};

File: packages/services/f-statistics/src/tests/index.test.js [hunk 2]
```diff
@@ -6,8 +7,116 @@ describe('f-statistics', () => {
...
-        expect(new StatisticsClient().publish).toBeDefined();
+            expect(new StatisticsClient().publish).toBeDefined();
+        });
+
+        it('should publish logs with expected properties', () => {
+            // Arrange
+            const statisticsClient = new StatisticsClient(justLogMock, null, null);
+
+            const expectedPayload = {
+                je_environment: 'test',
```
Comment (xander-marjoram) [inline packages/services/f-statistics/src/tests/index.test.js]: Nit: Looks like you could you use the `basePayload` here too?

File: packages/services/f-statistics/src/tests/index.test.js [hunk 3]
```diff
@@ -6,8 +7,116 @@ describe('f-statistics', () => {
...
+            // Assert
+            expect(publications[0]).toEqual(expectedPayload);
+        });
+
+        it('should publish logs with additional base payload properties when provided', () => {
+            // Arrange
+            const statisticsClient = new StatisticsClient(justLogMock, null, { a_base_payload_property: 'This is a test' });
+
+            const expectedPayload = {
+                je_environment: 'test',
```
Comment (xander-marjoram) [inline packages/services/f-statistics/src/tests/index.test.js]: And here

**Review by xander-marjoram — DISMISSED** [review]

File: packages/services/f-statistics/src/index.js [hunk 4]
```diff
@@ -2,22 +2,27 @@
...
-    publish (message, payload) {
+    publish (message, statisticPayload) {
         this.justLogInstance.info(message, {
-            je_feature: 'f-statistics',
-            je_feature_for: this.configuration?.featureName || 'Unspecified',
-            je_logType: 'client-stats',
-            je_environment: this.configuration?.environment || 'Unspecified',
-            ...payload
+            ...this.basePayload,
+            ...statisticPayload
```
Comment (minrwhite) [inline packages/services/f-statistics/src/index.js]: Are we okay with allowing keys from the basePayload (`je_feature` et al) to potentially be overridden here?

Comment (tommcclean) [reply packages/services/f-statistics/src/index.js]: I kind of don't mind it; but it makes sense to invert this as I already do this in the constructor with the base - so consistency I guess.

**Review by minrwhite — APPROVED** [review]

**Review by xander-marjoram — APPROVED** [review]

**Pull request merged**